libp2pdht/dht/providers_encoding.nim

```diff
@@ -34,12 +34,22 @@ func getField*(pb: ProtoBuffer, field: int,
     else:
       err(ProtoError.IncorrectBlob)
 
-func write*(pb: var ProtoBuffer, field: int, pr: PeerRecord) =
-  ## Write PeerRecord value ``pr`` to object ``pb`` using ProtoBuf's encoding.
-  write(pb, field, pr.encode())
+func write*[T: SignedPeerRecord | PeerRecord | Envelope](
+    pb: var ProtoBuffer,
+    field: int,
+    env: T) {.raises: [Defect, ResultError[CryptoError]].} =
+
+  ## Write Envelope value ``env`` to object ``pb`` using ProtoBuf's encoding.
+  let encoded = env.encode().tryGet()
+  write(pb, field, encoded)
+
+# TODO: This should be included upstream in libp2p/signed_envelope. Once it's
+# added in libp2p, we can remove it from here.
+proc encode*[T](msg: SignedPayload[T]): Result[seq[byte], CryptoError] =
```

Comment:
Done, thanks!

Was hesitant to also add the `write` above, but I'll keep out for now